Sprint 16

usvladis opened the pull request:
Проведена работа по 16 спринту:
-Экран со списком категорий переписан на архитектуру MVVM  с реализацией байндингов через замыкания. 
-Сделаны экраны онбординга с использованием UIPageViewController

usvladis committed:
Сделаны экраны онбординга с использованием UIPageViewController

usvladis committed:
С прошлым обновлением пропал Navigation Controller. Пофиксил

usvladis committed:
Экран со списком категорий переписан на архитектуру MVVM с реализацией байндингов через замыкания.

usvladis committed:
Поправил отображение хэдера

Helvetii951 commented on Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,107 @@
+import UIKit
+
+class OnboardingViewController: UIPageViewController {
+    lazy var pages: [UIViewController] = {
+        let red = RedPadeViewController()
+        let blue = BluePageViewController()
+        return [blue, red]
+    }()
```
`RedPadeViewController` по коду не отличается от `BluePageViewController`. Было бы классно сделать один общий класс для страницы "листалки" и настраивать, например, через инициализатор.

Helvetii951 replied:
`pages` можно сделать приватным. Рекомендую писать свойства и функции сразу с минимальным доступом. По необходимости менять.

Helvetii951 commented on Onboarding/BluePageViewController.swift:
```diff
@@ -0,0 +1,83 @@
+        return label
+    }()
+    
+    private let button: UIButton = {
+        let button = UIButton()
+        button.setTitle("Вот это технологии!", for: .normal)
+        button.tintColor = .white
```
Обычно делают одну кнопку на весь онбординг поверх скролла.

Helvetii951 commented on Onboarding/BluePageViewController.swift:
```diff
@@ -0,0 +1,83 @@
+class BluePageViewController: UIViewController {
+    
+    private let label: UILabel = {
+        let label = UILabel()
+        label.text = "Отслеживайте только то, что хотите"
+        label.font = UIFont(name: "YSDisplay-Bold", size: 32)
```
`label.text = "Отслеживайте только то, что хотите"`

Было бы классно обернуть инфу о странице(текст и картинку) в enum. [Сделать эту инфу вычислимыми свойствами](https://www.avanderlee.com/swift/computed-property/). Так код будет гибче.
```swift
enum PageModel {
  case somePage
  case anotherPage

  var image : UIImage { ... }
  var text : String { ... }
}

...

let secondPageVC = OnboardingSinglePageVC(pageModel : .anotherPage)
```

Helvetii951 commented on Onboarding/BluePageViewController.swift:
```diff
@@ -0,0 +1,83 @@
+        let newVC = TabBarViewController()
+        newVC.modalPresentationStyle = .fullScreen
+        present(newVC, animated: true, completion: nil)
+    }
```
Тут несколько рекомендаций:

1. Лучше забрать у контроллеров онбординга обязанность навигации.

Контроллер страницы сообщает о нажатии кнопки -> контейнер страниц сообщает об этом в `SceneDelegate` -> `SceneDelegate` уже решает что делать дальше.

2. Так контроллеры онбординга остаются в памяти. Лучше менять `rootViewController` главного окна.

Helvetii951 commented on Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,107 @@
+
+import UIKit
+
+class OnboardingViewController: UIPageViewController {
+    lazy var pages: [UIViewController] = {
+        let red = RedPadeViewController()
```
Рекомендую проставлять `final` везде, где это возможно. Преимущества:

1. Никто не сможет по ошибке сделать подкласс.
2. Присвоение классу статуса `final` указывает компилятору Swift, что метод следует вызывать напрямую (статическая диспетчеризация), а не искать функцию в таблице методов (динамическая диспетчеризация). Это снижает затраты на вызов функции и повышает производительность. Подробнее об этом можно прочитать [в блоге разработчика Swift](https://developer.apple.com/swift/blog/?id=27).

Helvetii951 commented on TackerViewController/TrackerViewController.swift:
```diff
@@ -16,112 +16,65 @@ final class TrackerViewController: UIViewController{
-    var visibleTrackers: [TrackerCategory] = []
-    var categories: [TrackerCategory] = [TrackerCategory(title: "Важное", trackers: [])]
-    var completedTrackers: [TrackerRecord] = []
+    private var viewModel: TrackerViewModel!
     
     override func viewDidLoad() {
         super.viewDidLoad()
+        viewModel = TrackerViewModel()
+        
```
Для гибкости и тестируемости кода желательно объявлять вью-модель протоколом и как-нибудь инжектить.

[Неплохая статья по внедрению зависимостей](https://habr.com/ru/companies/tinkoff/articles/546360/).

usvladis committed:
Экран онбординга отображается один раз, данные сохраняются в UserDefaults.
После удаления приложения при новом запуске экран онбординга напомнит пользователю о функционале приложения

usvladis committed:
Сверстаны экраны для выбора и добавления новых категорий(CategoriesViewController, NewCategoryViewController).
TODO: Добавить функциональность

usvladis committed:
Подготовил экраны добавления приввычек к отображению выбранной категории

usvladis committed:
Написаны экраны для добавления категорий
-Экраны написаны с использованием архитектуры MVVM

Доработан главный экран отображения трекеров и адаптирован под изменения.

TODO: Доработать дизайн.
На экране добавления нерегулярных привычек не отображается название выбранной категории

На экране категорий добавить закругление для TableView
Вернуть фон таблицы на экране Расписания

usvladis committed:
Проведена работа над ошибками в 16 спринте
